harness/api_checkers.go

```diff
@@ -0,0 +1,477 @@
+}
+
+// APICheckSuite is a testify suite for checking a Tracer against the OpenTracing API.
+type APICheckSuite struct {
+	suite.Suite
+	opts      APICheckCapabilities
+	newTracer func() (tracer opentracing.Tracer, closer func())
+	tracer    opentracing.Tracer
+	closer    func()
+}
+
+// NewAPICheckSuite returns a testify suite for checking a Tracer against the OpenTracing API.
+// It is provided a function that will be executed to create and destroy a tracer for each test
+// in the suite, and API test options described by APICheckCapabilities.
+func NewAPICheckSuite(
+	newTracer func() (tracer opentracing.Tracer, closer func()),
+	opts ...APICheckOption,
+) *APICheckSuite {
+	s := &APICheckSuite{newTracer: newTracer}
+	for _, o := range opts {
+		o.Apply(s)
+	}
+	return s
+}
+
+// APICheckOption instances may be passed to NewAPICheckSuite.
+type APICheckOption interface {
+	Apply(*APICheckSuite)
+}
+
+// Apply satisfies the APICheckOption interface.
+func (c APICheckCapabilities) Apply(s *APICheckSuite) {
+	s.opts = c
+}
+
+// CheckEverything enables all API checks.
+type CheckEverything struct{}
+
+// Apply satisfies the APICheckOption interface.
+func (CheckEverything) Apply(s *APICheckSuite) {
+	s.opts.CheckBaggageValues = true
+	s.opts.CheckExtract = true
+	s.opts.CheckInject = true
+}
+
+// UseProbe specifies an APICheckProbe implementation to use.
+type UseProbe struct {
+	APICheckProbe
+}
+
+// Apply satisfies the APICheckOption interface.
+func (u UseProbe) Apply(s *APICheckSuite) {
+	s.opts.Probe = u.APICheckProbe
+}
+
+// BeforeTest creates a tracer for this specific test invocation.
+func (s *APICheckSuite) BeforeTest(suiteName, testName string) {
+	s.tracer, s.closer = s.newTracer()
+	if s.tracer == nil {
+		panic(fmt.Sprintf("newTracer returned nil Tracer before running %s, %s", suiteName, testName))
+	}
+}
+
+// AfterTest closes the tracer, and clears the test-specific tracer.
+func (s *APICheckSuite) AfterTest(suiteName, testName string) {
+	if s.closer != nil {
+		s.closer()
+	}
+	s.tracer, s.closer = nil, nil
+}
+
+// TestStartSpan checks if a Tracer can start a span and calls some span API methods.
+func (s *APICheckSuite) TestStartSpan() {
+	span := s.tracer.StartSpan(
+		"Fry",
+		opentracing.Tag{Key: "birthday", Value: "August 14 1974"})
+	span.LogFields(
+		log.String("hospital", "Brooklyn Pre-Med Hospital"),
+		log.String("city", "Old New York"))
+	span.Finish()
+}
+
+// TestStartSpanWithParent checks if a Tracer can start a span with a specified parent.
+func (s *APICheckSuite) TestStartSpanWithParent() {
+	parentSpan := s.tracer.StartSpan("parent")
+	s.NotNil(parentSpan)
+
+	span := s.tracer.StartSpan(
+		"Leela",
+		opentracing.ChildOf(parentSpan.Context()))
+	span.Finish()
+
+	span = s.tracer.StartSpan(
+		"Leela",
+		opentracing.FollowsFrom(parentSpan.Context()),
+		opentracing.Tag{Key: "birthplace", Value: "sewers"})
+	if s.opts.Probe != nil {
+		s.True(s.opts.Probe.SameTrace(parentSpan, span))
```

Comment:
Nope, thanks for catching, I added another Probe.SameTrace check for the ChildOf span.